pkg/assembler/backends/testing/hasSourceAt.go

```diff
@@ -87,7 +86,129 @@ func registerAllHasSourceAt(client *demoClient) error {
+	if !srcHasNamespace {
+		return nil, gqlerror.Errorf("Source type \"%s\" not found", source.Type)
+	}
+	srcName, srcHasName := srcNamespace.namespaces[source.Namespace]
+	if !srcHasName {
+		return nil, gqlerror.Errorf("Source namespace \"%s\" not found", source.Namespace)
+	}
+	found := false
+	var sourceID nodeID
+	for _, src := range srcName.names {
+		if src.name != source.Name {
+			continue
+		}
+		if noMatchInput(source.Tag, src.tag) {
+			continue
+		}
+		if noMatchInput(source.Commit, src.commit) {
+			continue
+		}
+		if found {
+			return nil, gqlerror.Errorf("More than one source matches input")
+		}
+		sourceID = src.id
+		found = true
+	}
+	if !found {
+		return nil, gqlerror.Errorf("No source matches input")
+	}
+
+	pkgNamespace, pkgHasNamespace := packages[packageArg.Type]
+	if !pkgHasNamespace {
+		return nil, gqlerror.Errorf("Package type \"%s\" not found", packageArg.Type)
+	}
+	pkgName, pkgHasName := pkgNamespace.namespaces[nilToEmpty(packageArg.Namespace)]
+	if !pkgHasName {
+		return nil, gqlerror.Errorf("Package namespace \"%s\" not found", nilToEmpty(packageArg.Namespace))
+	}
+	pkgVersion, pkgHasVersion := pkgName.names[packageArg.Name]
+	if !pkgHasVersion {
+		return nil, gqlerror.Errorf("Package name \"%s\" not found", packageArg.Name)
+	}
+	var packageID nodeID
+	if pkgMatchType.Pkg == model.PkgMatchTypeAllVersions {
+		packageID = pkgVersion.id
+	} else {
+		found = false
+		for _, version := range pkgVersion.versions {
+			if noMatchInput(packageArg.Version, version.version) {
+				continue
+			}
+			if noMatchInput(packageArg.Subpath, version.subpath) {
+				continue
+			}
+			if !reflect.DeepEqual(version.qualifiers, getQualifiers(packageArg.Qualifiers)) {
+				continue
+			}
+			if found {
+				return nil, gqlerror.Errorf("More than one package matches input")
+			}
+			packageID = version.id
+			found = true
+		}
+		if !found {
+			return nil, gqlerror.Errorf("No package matches input")
+		}
+	}
+
+	// store the link
+	newSrcMapLink := &srcMapLink{
+		id:            c.getNextID(),
+		sourceID:      sourceID,
+		packageID:     packageID,
+		knownSince:    hasSourceAt.KnownSince.UTC(),
+		justification: hasSourceAt.Justification,
+		origin:        hasSourceAt.Origin,
+		collector:     hasSourceAt.Collector,
+	}
+	index[newSrcMapLink.id] = newSrcMapLink
+	sourceMaps = append(sourceMaps, newSrcMapLink)
+	// set the backlinks
+	index[packageID].(pkgNameOrVersion).setSrcMapLink(newSrcMapLink.id)
+	index[sourceID].(*srcNameNode).srcMapLink = newSrcMapLink.id
+
+	// build return GraphQL type
+	p, err := buildPackageResponse(packageID, nil)
+	if err != nil {
+		return nil, err
+	}
+	s, err := buildSourceResponse(sourceID, nil)
+	if err != nil {
+		return nil, err
+	}
+	out := model.HasSourceAt{
+		Package:       p,
+		Source:        s,
+		KnownSince:    hasSourceAt.KnownSince.UTC(),
+		Justification: hasSourceAt.Justification,
+		Origin:        hasSourceAt.Origin,
+		Collector:     hasSourceAt.Collector,
```

Comment:
Very minor nit: We can take these fields from `newSrcMapLink` instead of from input; I think it makes it clearer that the return is built from the thing we store